### PR TITLE
fix: Disable all emails per event type w/o a workflow [CAL-4066]

### LINKED ISSUE #12920: [CAL-3416] Ability to turn the cal.com email flow off
### Is your proposal related to a problem?

Currently, users automatically receive cal.com automated emails when they book events. This is amazing in most use cases. However, I have a few event types where we would want to send more custom emails that include further information. 

### Describe the solution you'd like

Combining webhooks and the ability to turn off the cal.com booking flow emails would allow us to build any email notification workflow we want.

### Describe alternatives you've considered

I haven't found any way to stop the cal.com automated emails.

##### House rules
- If this issue has a `🚨 needs approval` label, don't start coding yet. Wait until a core member approves feature request by removing this label, then you can start coding.
  - For clarity: Non-core member issues automatically get the `🚨 needs approval` label.
  - Your feature ideas are invaluable to us! However, they undergo review to ensure alignment with the product's direction.

<sub>[CAL-3416](https://linear.app/calcom/issue/CAL-3416/ability-to-turn-the-calcom-email-flow-off)</sub>

### PR DESCRIPTION
**Pull request by joeauyeung**
## What does this PR do?

This PR enables team event types under an org to disable the standard emails without having a workflow set.

There was also some confusion with some users where they believed that the current option would disable all emails. This PR expands disabling emails for the following:
- Scheduled
- Requested bookings
- Request reminders
- Requested denied
- Reschedule
- Requested reschedule
- Cancelled
- Payment requested
- Round robin emails to team members
- Seats specific emails

Before we would only check if we should send the email before calling the email functions. Now we send the whole event type metadata to the email function and have the check logic in the function.

This also introduces a new confirmation if the user wants to disable emails
CleanShot 2024-07-24 at 23 09 47@2x

- Fixes #12920
- Fixes CAL-4066 (Linear issue number - should be visible at the bottom of the GitHub issue description)

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Under an event type in a team in an org you should be able to see the two options under the advanced tab
- When enabling the options a dialog should appear where you have to type confirm to enable the option
- Enable both options
- Under each type of email that can be disabled, create a booking to test that the emails ARE NOT SENT
- Disable the option and ensure emails ARE SENT

### CODE REVIEW

**Review by joeauyeung — COMMENTED** [review]
I would look at `email-manager.ts` to see which emails are being affected by this PR.

When getting the metadata there is a lot of optional chaining when getting the event type metadata. This is because dynamic bookings might not have an event type associated with them in the DB. 

I have omitted optional chaining where an event type is guaranteed such as bookings that require confirmation and payments associated with them.

**Review by CarinaWolli — COMMENTED** [review]
Early review: 

- We said to allow disabling all emails without workflows only for orgs. But now every user can disable ALL emails only by having one workflow that triggers when booking is created.  
  - we either allow it for all users to disable emails without workflows 
  - or, non-org users can only disable the booking confirmation email (not all emails) when the workflows exists --> same as it was before 
- What happens to event types that already had the confirmation email disabled before? To me it looks like this would now change for them and all emails would stop sending
- Orgs can only disable the emails for team event types. I think we want that for personal event types as well

File: apps/web/components/eventtype/EventAdvancedTab.tsx [hunk 1]
```diff
@@ -569,6 +573,34 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupProps,
           )}
         />
       )}
+      {hasValidLicense && (
```
Comment (joeauyeung) [inline apps/web/components/eventtype/EventAdvancedTab.tsx:576]: We now only allow these options for enterprise customers. We'll still keep both options available but will replace once we allow users to set granular options for emails
CleanShot 2024-07-25 at 15 18 50@2x

File: packages/prisma/zod-utils.ts [hunk 2]
```diff
@@ -81,6 +81,12 @@ export const EventTypeMetaDataSchema = z
     disableSuccessPage: z.boolean().optional(),
     disableStandardEmails: z
       .object({
+        all: z
+          .object({
+            host: z.boolean().optional(),
+            attendee: z.boolean().optional(),
+          })
+          .optional(),
```
Comment (joeauyeung) [inline packages/prisma/zod-utils.ts:89]: We now have a new option to disable all emails while still keeping the old option for just confirmation emails.

**Review by joeauyeung — COMMENTED** [review]
@CarinaWolli addressed your feedback.

I don't think there would be a lot of use cases where an individual wants to disable all emails as it takes a lot of setup which might be specific to their company. If a company did want this for individual events they should use the managed event type.

File: apps/web/components/eventtype/EventAdvancedTab.tsx [hunk 3]
```diff
@@ -569,6 +573,34 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupProps,
...
+      {team?.parentId && (
+        <Controller
+          name="metadata.disableStandardEmails.all.host"
+          defaultValue={!!formMethods.getValues("seatsPerTimeSlot")}
+          render={({ field: { value, onChange } }) => (
+            <>
+              <DisableAllEmailsSetting checked={value} onCheckedChange={onChange} recipient="hosts" t={t} />
+            </>
+          )}
+        />
```
Comment (CarinaWolli) [inline apps/web/components/eventtype/EventAdvancedTab.tsx:602]: I think we can move this up right after themetadata.disableStandardEmails.all.attendee Controller. No need check for `team?.parentId` again

File: apps/web/components/eventtype/EventAdvancedTab.tsx [hunk 4]
```diff
@@ -56,6 +58,8 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupProps,
   );
   const [hashedUrl, setHashedUrl] = useState(eventType.hashedLink?.link);
   const bookingFields: Prisma.JsonObject = {};
+  const session = useSession();
+  const hasValidLicense = session.data ? session.data.hasValidLicense : null;
```
Comment (CarinaWolli) [inline apps/web/components/eventtype/EventAdvancedTab.tsx:62]: `hasValidLicense` isn't used

File: apps/web/components/eventtype/settings/DisableAllEmailsSetting.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    recipient === "attendees" ? t("disable_all_emails_to_attendees") : t("disable_all_emails_to_hosts");
+
+  return (
+    <div>
+      <Dialog open={dialogOpen} onOpenChange={(e) => setDialogOpen(e)}>
+        <DialogContent title={title} Icon="circle-alert">
+          <p className="text-default text-sm">
+            <Trans i18nKey="disable_attendees_emails_description">
+              This will disable all emails to {{ recipient }}. This includes booking confirmations, requests,
+              reschedules and reschedule requests, and cancellation emails.
```
Comment (CarinaWolli) [inline apps/web/components/eventtype/settings/DisableAllEmailsSetting.tsx:41]: We are missing the location change emails here (maybe some other email too)

Should we add something like 'and any other emails related to booking updates'

File: apps/web/components/eventtype/settings/DisableAllEmailsSetting.tsx [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+              <br />
+              <br />
+              It is your responsibility to ensure that your {{ recipient }} are aware of any bookings and
+              changes to their booking.
```
Comment (CarinaWolli) [inline apps/web/components/eventtype/settings/DisableAllEmailsSetting.tsx:45]:
```suggestion
              changes to their bookings.
```

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
- [ ]  `sendScheduledEmails` and `sendScheduledSeatsEmails` doesn't support disabling all emails (only the old feature to disable confirmation email when workflow enabled is supported)
- [ ] In `eventTypes/update.handler.ts` we need to check if the event type is from an org before updating `metadata.disableStandardEmails.all`
- [ ] we also should not send reminder emails in`scheduleMandatoryReminder` if all emails to attendees are disabled

**Comment by joeauyeung** [comment]
> * [ ]   `sendScheduledEmails` and `sendScheduledSeatsEmails` doesn't support disabling all emails (only the old feature to disable confirmation email when workflow enabled is supported)
> * [ ]  In `eventTypes/update.handler.ts` we need to check if the event type is from an org before updating `metadata.disableStandardEmails.all`
> * [ ]  we also should not send reminder emails in`scheduleMandatoryReminder` if all emails to attendees are disabled

Addressed all feedback

File: packages/trpc/server/routers/viewer/eventTypes/update.handler.ts [hunk 7]
```diff
@@ -255,9 +259,15 @@ export const updateHandler = async ({ ctx, input }: UpdateOptions) => {
     };
   }
 
-  if (input.metadata?.disableStandardEmails) {
-    //check if user is allowed to disabled standard emails
+  if (input.metadata?.disableStandardEmails?.all) {
+    if (!eventType.team.parentId) {
```
Comment (CarinaWolli) [inline packages/trpc/server/routers/viewer/eventTypes/update.handler.ts:263]: type error here `'eventType.team' is possibly 'null'`

File: packages/emails/email-manager.ts [hunk 8]
```diff
@@ -201,7 +242,7 @@ export const sendScheduledSeatsEmails = async (
 
   const emailsToSend: Promise<unknown>[] = [];
 
-  if (!hostEmailDisabled) {
+  if (!hostEmailDisabled && !eventTypeDisableHostEmail(eventTypeMetadata)) {
```
Comment (CarinaWolli) [inline packages/emails/email-manager.ts:245]: `eventTypeMetadata` is missing as a param

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
Found some places where we don't pass `eventTypeMetadata` when calling the email function, let me know in case it's not needed there:
- [ ] `sendRoundRobinCancelledEmails` in new `roundRobinReassignment` function 
- [ ] `sendCancelledEmails` in `handleDeleteCredential`
- [ ] `sendRequestRescheduleEmail` in `app-store/vital/lib/reschedule.ts`

**Review by CarinaWolli — APPROVED** [review]
All my comments have been addressed, looks good now 🙏

**Comment by YPC-Avishai** [comment]
Is this feature available for teams on cal.com, too? 
Because I can not find those toggles. @joeauyeung

**Comment by janrabek** [comment]
I can confirm @YPC-Avishai

We really need this feature, does anyone know when it's going to be released on cal.com?

**Comment by nakul-malhotra** [comment]
is this ever going to be introduced for individuals? why are we blocking this to enterprise only?

**Pull request merged**